common/datablocks/src/kernels/data_block_group_by_hash.rs

```diff
@@ -273,16 +296,118 @@ where
+                    group_columns,
+                    ptr,
+                    step,
+                    &mut null_part_offset,
+                )?;
+                size /= 2;
+            }
+        } else {
+            while size > 0 {
+                build(size, &mut offsize, group_columns, ptr, step)?;
+                size /= 2;
+            }
         }
         Ok(group_keys)
     }
+
+    fn group_by_get_indices(
+        &self,
+        block: &DataBlock,
+        column_names: &[String],
+    ) -> Result<GroupIndices<Self::HashKey>> {
+        // Table for <group_key, (indices, keys) >
+        let mut group_indices = GroupIndices::<Self::HashKey>::default();
+        // 1. Get group by columns.
+        let mut group_columns = Vec::with_capacity(column_names.len());
+        {
+            for col in column_names {
+                group_columns.push(block.try_column_by_name(col)?);
+            }
+        }
+
+        // 2. Build serialized keys
+        let group_keys = self.build_keys(&group_columns, block.num_rows())?;
+        // 2. Make group with indices.
+        {
+            for (row, group_key) in group_keys.iter().enumerate().take(block.num_rows()) {
+                match group_indices.get_mut(group_key) {
+                    None => {
+                        let mut group_values = Vec::with_capacity(group_columns.len());
+                        for col in &group_columns {
+                            group_values.push(col.get(row));
+                        }
+                        group_indices.insert(*group_key, (vec![row as u32], group_values));
+                    }
+                    Some((v, _)) => {
+                        v.push(row as u32);
+                    }
+                }
+            }
+        }
+
+        Ok(group_indices)
+    }
+
+    fn group_by(
+        &self,
+        block: &DataBlock,
+        column_names: &[String],
+    ) -> Result<GroupBlock<Self::HashKey>> {
+        let group_indices = self.group_by_get_indices(block, column_names)?;
+        // Table for <(group_key, keys, block)>
+        let mut group_blocks = GroupBlock::<Self::HashKey>::with_capacity(group_indices.len());
+
+        for (group_key, (group_indices, group_keys)) in group_indices {
+            let take_block = DataBlock::block_take_by_indices(block, &group_indices)?;
+            group_blocks.push((group_key, group_keys, take_block));
+        }
+
+        Ok(group_blocks)
+    }
+}
+
+/// Init the nullable part's offset in bytes. It follows the values part.
+#[inline]
+fn init_nullable_offset(null_part_offset: &mut usize, group_keys: &[&ColumnRef]) -> Result<()> {
+    for group_key_column in group_keys {
+        *null_part_offset += remove_nullable(&group_key_column.data_type())
+            .data_type_id()
+            .numeric_byte_size()?;
+    }
+    Ok(())
+}
+
+/// Init the nullable part's offset in bytes. It follows the values part.
+#[inline]
+fn init_nullable_offset_via_fields(
```

Comment:
why not make this function return `Result<usize>`

Reply:
> why not make this function return `Result<usize>`

I see some codes use this kind of initialization as c style.  Using Result<> makes codes more healthy, it is a better way.